### PR TITLE
Add BigramStringEmission class

### PR DESCRIPTION
**Pull request by ThomasColthurst**
Also, fix a bug in string_alignment::topk_alignments where it would create alignment pieces past the end of one of the strings.

### CODE REVIEW

File: cxx/emissions/bigram_string.cc [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+    const std::pair<std::string, std::string>& x) const {
+  // Store state.
+  double old_N = N;
+  std::vector<DirichletCategorical> orig_insertions;
+  std::vector<DirichletCategorical> orig_substitutions;
+  std::copy(insertions.begin(), insertions.end(), std::back_inserter(orig_insertions));
+  std::copy(substitutions.begin(), substitutions.end(),
+            std::back_inserter(orig_substitutions));
+  double orig_lp = logp_score();
+  BigramStringEmission* unsafe_this = const_cast<BigramStringEmission*>(this);
```
Comment (emilyfertig) [inline cxx/emissions/bigram_string.cc:185]: Can we avoid this, maybe by enabling the copy constructor and modifying a copy instead? Or by declaring insertions/substitutions `mutable`, similar to the bigram distribution?

**Review by emilyfertig — COMMENTED** [review]
Nice!

File: cxx/emissions/bigram_string.cc (hunk 1 above)
Comment (ThomasColthurst) [reply cxx/emissions/bigram_string.cc]: Excellent idea!  I enabled the copy constructor and now it is so much nicer!!

**Review by emilyfertig — APPROVED** [review]
Agreed, LG!

**Pull request merged**